lib/Sema/CSSimplify.cpp

```diff
@@ -6080,10 +6090,33 @@ ConstraintSystem::simplifyApplicableFnConstraint(
   ConstraintLocatorBuilder outerLocator =
     getConstraintLocator(anchor, parts, locator.getSummaryFlags());
 
-  // Before stripping optional types, save original type for handling
-  // @dynamicCallable applications. This supports the fringe case where
-  // `Optional` itself is extended with @dynamicCallable functionality.
-  auto origType2 = desugar2;
+  // Handle applications of types with `callAsFunction` methods.
+  // Do this before stripping optional types below, when `shouldAttemptFixes()`
+  // is true.
+  auto hasCallAsFunctionMethods =
+      desugar2->mayHaveMembers() &&
+      llvm::any_of(lookupMember(desugar2, DeclName(ctx.Id_callAsFunction)),
+                   [](LookupResultEntry entry) {
+                     return isa<FuncDecl>(entry.getValueDecl());
+                   });
+  if (hasCallAsFunctionMethods) {
+    auto memberLoc = getConstraintLocator(
+        outerLocator.withPathElement(ConstraintLocator::Member));
+    // Add a `callAsFunction` member constraint, binding the member type to a
+    // type variable.
+    auto memberTy = createTypeVariable(memberLoc, TVO_CanBindToLValue |
+                                                  TVO_CanBindToNoEscape |
+                                                  TVO_CanBindToInOut);
```

Comment:
I don't think member type could be bound to any of these types listed in flags btw since it represents a result type of the member choice, `inout` makes sense only to arguments same way as `l-value` and can't really return no-escape function right?

Reply:
I see, that makes sense. How can I call `createTypeVariable` with no flag arguments?

```
swift/lib/Sema/CSSimplify.cpp:6108:49: error: too few arguments to function call, expected 2, have 1
    auto memberTy = createTypeVariable(memberLoc);
                    ~~~~~~~~~~~~~~~~~~          ^
```

Reply:
`createTypeVariable(memberLoc, /*options=*/0);

Reply:
Done in https://github.com/apple/swift/pull/24299/commits/cebbd6335dab7e0bbf3e5c0dff9c445588e01f81.